formsfeeder.pf4j-spring/src/main/java/com/_4point/aem/formsfeeder/pf4j/spring/EnvironmentConsumer.java

```diff
@@ -11,6 +11,9 @@
 	public static final String FORMSFEEDER_PLUGINS_ENV_PARAM_PREFIX = "formsfeeder.plugins.";
 	public static final String AEM_HOST_ENV_PARAM = FORMSFEEDER_PLUGINS_ENV_PARAM_PREFIX + "aemHost";
 	public static final String AEM_PORT_ENV_PARAM = FORMSFEEDER_PLUGINS_ENV_PARAM_PREFIX + "aemPort";
+	public static final String AEM_USE_SSL_PARAM = FORMSFEEDER_PLUGINS_ENV_PARAM_PREFIX + "aemUseSsl";
+	public static final String AEM_USERNAME_PARAM = FORMSFEEDER_PLUGINS_ENV_PARAM_PREFIX + "aemUsername";
+	public static final String AEM_SECRET_PARAM = FORMSFEEDER_PLUGINS_ENV_PARAM_PREFIX + "aemSecret";

```

Comment:
Just a question:  These all correspond to parameters already in the FormsFeeder client interface except for aemSecret.  I take aemSecret is for the API key (per the PR #2 )?   You've chosen the word "secret" so that it could be used for any of the following: a password, an apiKey or an encrypted password?

I like the idea of being less specific, so "secret" works for me, I just want to make sure that I'm liking it for the right reasons. :)

Reply:
First thing, I think these parameters apply to setting up the call to AEM instance (so already after the formsfeeder entry point, not related to FormsFeeder endpoint).

TraditionalFormsService formsAdapter = RestServicesFormsServiceAdapter.builder()
    .machineName(aemHostName())
    .port(aemHostPort())
    .basicAuthentication(aemUsername(), aemPassword()) // default to admin/admin in Example plugin but our cloud instance uses different credentials than admin admin, we want to get them from Environment (application.properties file) rather than each plugin dealing with them individually
    .useSsl(aemUseSsl()) //if aemHostName is not https, this has to be false. if it is https as in the cloud endpoints, it has to be true, our on prem instance we started out with was http and our cloud instances are https so we wanted to be able to filp this value related to what we flipped hostName to.
					.build();

Then I think maybe our changes to FormsFeeder.client are related to calling FormsFeeder.service via an api gateway rather than calling the formsfeeder directly. So we aren't calling https://ourformsfeeder.server.com/api/v1/Plugin. We are calling https://ourapigateway.whatever.com/somecontext/whatever/Plugin?apikey=abc. The api gateway handles our service contracts checking apikey make sure auth is good.. then forwards call to our physical formsfeeder service endpoint https://ourformsfeeder.server.com/api/v1/Plugin endpoint.

This is where I'm wondering if maybe instead of the flexibility built into FormsFeeder.client to basically allow hitting some arbitrary endpoint (the apigateway) to then be forwarded to FormsFeeder.service, we may need to just implement our own specific client with this stuff, leaving the project implemented client specific to hitting the service directly?

Reply:
> First thing, I think these parameters apply to setting up the call to AEM instance (so already after the formsfeeder entry point, not related to FormsFeeder endpoint).

You're absolutely right, I misspoke.  I should have said rest-services.client Builder.

> This is where I'm wondering if maybe instead of the flexibility built into FormsFeeder.client to basically allow hitting some arbitrary endpoint (the apigateway) to then be forwarded to FormsFeeder.service, we may need to just implement our own specific client with this stuff, leaving the project implemented client specific to hitting the service directly?

That's certainly a possibility however I don't think what you're looking for is that uncommon or unique.  I'd just like to see things implemented in a general way so that they apply to your project and potentially to other projects.    I think my response in PR #2 heads in that direction.  I think that adding HTTP headers or query parameters to a request are general facilities that could be incorporated into the codebase that would allow you to perform the specific operations you need.

>  https://ourapigateway.whatever.com/somecontext/whatever/Plugin?apikey=abc.

On a side note, my read of the PR #2 code has the Authentication key being passed in an HTTP header.  Is this the same as the API key being passed as a query parameter above or are these two different parameters?  If they are the same, does it need to be a query parameter or a http header?  (I'm just looking to understand whether the need is to implement HTTP headers, query parameters or both.)

Reply:
With respect to the custom URL (i,e, PR #2), I don't have a problem with the way this is implemented in that PR, so I think that with that, plus the implementation of custom HTTP headers and/or custom query parameters, that should cover your use case.  Let me know if you don't think this is the case.

Reply:
I think these are two separate things altogether. #2 is attempting to solve for hitting the formsfeeder service via an apigateway with apikey authentication. This one is solving for how to connect to AEM instance within plugins.

The aemUseSsl, aemPassword, and aemUsername ARE already available on the rest-services.client that's true. However, when the Plugins instantiate those Client objects, they need to get the values used to set those from somewhere, and that's what I'm adding, but I want to add them as standard environment based properties so that the constant of the property value is the same across all of our plugins and not hardcoded as the same by chance.

For Example from Example plugin we get aemHostName value to set on the rest-services.client with this code. This retrieves the value from application.properties property formsfeeder.plugins.aemHostName=localhost for example.
```java
		private String aemHostName() {
			if (this.aemHostName == null) {
				this.aemHostName = Objects.requireNonNull(environment, "Plug-in's Environment has not been populated!").getRequiredProperty(EnvironmentConsumer.AEM_HOST_ENV_PARAM);
			}
			return this.aemHostName;
		}
```
and the password is simply hardcoded. 
```java
private String aemPassword() {
			if (this.aemPassword == null) {
				this.aemPassword = "admin";		// Hardcoded for now, may change this later 
			}
			return this.aemPassword;
		}
```
So we are currently adding a private String constant to EACH plugin that matches the property name, formsfeeder.plugins.aemPassword (well secret but I think that should change to password now) = admin or our non-admin password for cloud instance. So given those properties are coming from the same place I thought it appropriate that the property name be in the EnvironmentConsumer interface as a constant and then retrieved thusly:
```java
	private String aemPassword() {
		if (this.aemPassword == null) {
			this.aemPassword = Objects.requireNonNull(environment, "Plug-in's Environment has not been populated!").getProperty(AEM_SECRET_PARAM, "admin");
		}
		return this.aemPassword;
	}
```

Reply:
Which means, I should also update FormsFeeder.plugins to use the same logic, but that's all part of that Test cases that haven't yet been written I would suppose..